lk2nd/smp/cpu-boot.c

```diff
@@ -56,6 +130,79 @@ int qcom_set_boot_addr(uint32_t addr, bool arm64)
 				QCOM_SCM_BOOT_FLAG_COLD_ALL, addr);
 }
 
+void qcom_power_up_l2_cache(uint32_t mpidr, uint32_t base)
+{
+	uint32_t val;
+
+	/* Skip if cluster L2 is already powered on */
+	val = readl(base + APCS_L2_PWR_STATUS_ADDR);
+	if (val & BIT(9)) {
+		dprintf(INFO, "CPU (%d) L2 cache already powered-up\n", mpidr);
+		return;
+	}
+
+	/*
+	 * Close L2/SCU Logic GDHS and power up the cache
+	 *
+	 * Downstream value write value 0x10D700 decodes to
+	 * 100001101011100000000 including undocumented BIT(15) | BIT(14)
+	 * 20 10000	APCS_L2_HS_CNT(CNT) => 0x10
+	 * 15 1		APCS_L2_SCU_MAGIC_BIT1
+	 * 14 10	APCS_L2_SCU_MAGIC_BIT0
+	 * 12 10	APCS_L2_PWR_CTL_SYS_RESET
+	 * 10 1		APCS_L2_PWR_CTL_L2_HS_RST
+	 * 9  1		APCS_L2_PWR_CTL_L2_HS_EN
+	 * 8  1		APCS_L2_PWR_CTL_L2_HS_CLAMP
+	 */
+	val = APCS_L2_HS_CNT(0x10) | APCS_L2_SCU_MAGIC_BIT1 |
```

Comment:
Another very minor nit: I usually count using decimals, not hex. This is a bit like writing `udelay(0x36)` instead of `udelay(54)` otherwise. ;)

```suggestion
	val = APCS_L2_HS_CNT(16) | APCS_L2_SCU_MAGIC_BIT1 |
```

Reply:
Sure decimal it is

Reply:
I find it a bit easier to mentally calculate shifts in hex ..

Reply:
Right, from that perspective hex would be better.

My thinking comes from an (idealized) standpoint where we would actually understand the hardware well enough to write such a sequence on our own (rather than just decoding the sequence from Qualcomm). In that case it does not matter if it ends up being 0x10 shifted, 0x42 or whatever. We want to set the "count value" to 16 and the compiler can worry about assembling the register value. :)